### PR TITLE
Fix merging artifact chunks error when minio storage basepath is set

### PR DESCRIPTION
**Pull request by fuxiaohei**
Related to  https://github.com/go-gitea/gitea/issues/28279

When merging artifact chunks, it lists chunks from storage. When storage is minio, chunk's path contains `MINIO_BASE_PATH` that makes merging break.

<del>So trim the `MINIO_BASE_PATH` when handle chunks.</del>

Update the chunk file's basename to retain necessary information. It ensures that the directory in the chunk's path remains unaffected.

### CODE REVIEW

File: routers/api/actions/artifacts_chunks.go [hunk 1]
```diff
@@ -64,10 +67,16 @@ type chunkFileItem struct {
...
+func listChunksByRunID(st storage.ObjectStorage, runID int64, basepath string) (map[int64][]*chunkFileItem, error) {
 	storageDir := fmt.Sprintf("tmp%d", runID)
 	var chunks []*chunkFileItem
 	if err := st.IterateObjects(storageDir, func(path string, obj storage.Object) error {
+		// if basepath is settings.Actions.ArtifactStorage.MinioConfig.BasePath, it should trim it
+		if basepath != "" {
+			log.Debug("listChunksByRunID, trim basepath: %s, path: %s", basepath, path)
+			basepath = strings.TrimPrefix(basepath, "/")
+			path = strings.TrimPrefix(path, basepath+"/")
+		}
```
Comment (wxiaoguang) [inline routers/api/actions/artifacts_chunks.go]: It's too hacky. Why not extract the "base name" of the file and parse it?

I do not see why it makes sense to put the dir into Sscanf's string `fmt.Sscanf(path, filepath.Join(storageDir ...))`

Comment (fuxiaohei) [reply routers/api/actions/artifacts_chunks.go]: In current design, the artifact chunk save as `tmp{run-id}/{art-id}-{start}-{end}.chunk`. Make different run chunks in different dir. 

I think I can change it to `tmp{run-id}/tmp-{run-id}-{art-id}-{start}-{end}.chunk`. The basename of file is enough to scan.

File: routers/api/actions/artifacts_chunks.go [hunk 2]
```diff
@@ -67,9 +70,12 @@ type chunkFileItem struct {
 func listChunksByRunID(st storage.ObjectStorage, runID int64) (map[int64][]*chunkFileItem, error) {
 	storageDir := fmt.Sprintf("tmp%d", runID)
 	var chunks []*chunkFileItem
-	if err := st.IterateObjects(storageDir, func(path string, obj storage.Object) error {
-		item := chunkFileItem{Path: path}
-		if _, err := fmt.Sscanf(path, filepath.Join(storageDir, "%d-%d-%d.chunk"), &item.ArtifactID, &item.Start, &item.End); err != nil {
+	if err := st.IterateObjects(storageDir, func(fpath string, obj storage.Object) error {
+		baseName := filepath.Base(fpath)
```
Comment (lunny) [inline routers/api/actions/artifacts_chunks.go:73]: You should trim MINIO_BASE_PATH ?

Comment (fuxiaohei) [reply routers/api/actions/artifacts_chunks.go:73]: No need now. Use `basename` of the file to fscan. No matter the parent directory path.

**Review by delvh — APPROVED** [review]
Took me a bit to understand the fix, but I understand it now.

However, what really confuses me: Aren't PR description and fix completely different?

File: routers/api/actions/artifacts_chunks.go [hunk 3]
```diff
@@ -67,9 +70,12 @@ type chunkFileItem struct {
...
+		// when read chunks from storage, it only contains storage dir and basename,
+		// no matter the subdirectory setting in storage config
+		item := chunkFileItem{Path: path.Join(storageDir, baseName)}
```
Comment (wxiaoguang) [inline routers/api/actions/artifacts_chunks.go]: It confused me that it sometimes calls `filepath` (line 74) while sometimes calls `path` (here). It looks quite fragile to me.

At least, IMO it needs to normalize the "paths" before using it.

Comment (fuxiaohei) [reply routers/api/actions/artifacts_chunks.go]: The second `path.Join` is not necessary. It saves chunk as filename `tmp/xxx.chunk` in above. So just `storageDir +"/" + basename` is enough.

Comment (wxiaoguang) [reply routers/api/actions/artifacts_chunks.go]: I still have difficulty to understand whether it is right to do these magic path operations.

Could it clarify these "path kinds"? If a path is a filesystem path, then it needs `filepath` package, if it is a web path (or a web path converted from filepath), then it needs `util.PathJoinRelX`.

Comment (fuxiaohei) [reply routers/api/actions/artifacts_chunks.go]: `baseName := filepath.Base(fpath)` is the real fix. `fpath` is from Storage and  can be filesystem path so use `filepath` to get basename.  

`chunkFileItem.Path` is a key to read and write file for Storage. When saving chunk, it uses `fmt.Sprintf("tmp%d/%d-%d-%d-%d.chunk", runID, runID, artifact.ID, start, end)` with `/` slash when saving into Storage. 

Using `storageDir +"/" + basename` is enough in item.Path to read from Storage. Storage module can handle os difference.

**Review by lunny — APPROVED** [review]

**Pull request merged**